### PR TITLE
Adding configuration for redirecting journalctl logs to a file

### PR DESCRIPTION
**Pull request by YashviJain01**
### :nut_and_bolt: Description: What code changed, and why?

After adding the below configuration the logs will be redirecting to /var/log/automate.log

[global.v1.log]
   redirect_sys_log = true
redirect_log_file_path = "/var/tmp/"
compress_rotated_logs = true
max_size_rotate_logs = "10M"
max_number_rotated_logs = 2

### :chains: Related Resources

https://chefio.atlassian.net/browse/STALWART-231
https://chefio.atlassian.net/browse/STALWART-232
https://chefio.atlassian.net/browse/STALWART-227

### :+1: Definition of Done

Whenever users patches this command a file named automate.log will be formed in /var/log/ folder. Which will have logrotate policies enabled.

if redirect_sys_log is set to be true, it will give an error if redirect_log_file_path is not having any value. That is its a required parameter
The default value is as follows for others:

compress_rotated_logs =false
max_size_rotate_logs = "100M"
max_number_rotated_logs = 10

### :athletic_shoe: How to Build and Test the Change

rebuild components/automate-deployment/
rebuild components/automate-cli/

Patch the below file :

[global.v1.log]
redirect_sys_log = true
compress_rotated_logs = true
max_size_rotate_logs = "10M"
max_number_rotated_logs = 2

### :white_check_mark: Checklist

**All PRs** must tick these:

- [X] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [X] All commits signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

With occasional exceptions, all PRs **from Progress employees** must tick these:

- [X] Is the code clear? *(complicated code or lots of comments--subdivide and use well-named methods, meaningful variable names, etc.)*
- [X] Consistency checked? *(user notifications, user prompts, visual patterns, code patterns, variable names)*
- [X] Repeated code blocks eliminated? *(adapt and reuse existing components, blocks, functions, etc.)*
- [X] Spelling, grammar, typos checked? *(at a minimum use `make spell` in any component directory)*
- [X] Code well-formatted? *(indents, line breaks, etc. improve rather than hinder readability)*

All PRs **from Progress employees** should tick these if appropriate:

- [X] Tests added/updated? (all new code needs new tests)
- [X] Docs added/updated? (all customer-facing changes)

*Please add a note next to any checkbox above if you are NOT ticking it.*

### CODE REVIEW

**Comment by Dmaddu** [comment]
@YashviJain01  @daveaugustus  Please record a demo video and attach it to this PR.

**Comment by YashviJain01** [comment]
https://user-images.githubusercontent.com/91940132/199649469-87579352-955f-49ef-979c-cb6f05bd0394.mov

**Comment by YashviJain01** [comment]
https://user-images.githubusercontent.com/91940132/199951865-387c6f8c-d4f5-4171-aaa1-5410c68fd247.mov

**Review by kalroy — APPROVED** [review]

File: components/automate-deployment/pkg/server/helpers.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+	buffer.WriteString(path + " {" + "\n")
+	for _, param := range params {
+		buffer.WriteString("\t" + param + "\n")
+	}
+	buffer.WriteString("}" + "\n")
+	return buffer.String()
+}
+
+func getLogFileName(path string) string {
+	return fmt.Sprintf("%sautomate.log", path)
```
Comment (Dmaddu) [inline components/automate-deployment/pkg/server/helpers.go:22]: there can be a possibility that the user doesn't provide the `/` at the end. So please make sure to support that as well.

File: components/automate-deployment/pkg/server/automate_config.go [hunk 2]
```diff
@@ -235,3 +254,139 @@ func (s *server) updateUserOverrideConfigFromRestoreBackupRequest(req *api.Resto
...
+
+func setConfigForRedirectLogs(req *api.PatchAutomateConfigRequest, existingCopy *deployment.AutomateConfig) error {
+
+	//Set the config if already not set
+	if req.GetConfig().GetGlobal().GetV1().GetLog().GetRedirectSysLog().GetValue() == true &&
+		existingCopy.GetGlobal().GetV1().GetLog().GetRedirectSysLog().GetValue() == false {
+
+		err := createConfigFileForAutomateSysLog(req.GetConfig().GetGlobal().GetV1().GetLog().GetRedirectLogFilePath().GetValue())
+		if err != nil {
+			return status.Error(codes.Internal, "Failed to create configuration into syslog")
```
Comment (Dmaddu) [inline components/automate-deployment/pkg/server/automate_config.go:266]: here we are swallowing the actual error. If we would like to display an error message similar to this, then write the actual error message at least in the logs.

File: components/automate-deployment/pkg/server/automate_config.go [hunk 3]
```diff
@@ -235,3 +254,139 @@ func (s *server) updateUserOverrideConfigFromRestoreBackupRequest(req *api.Resto
...
+		}
+		err = restartSyslogService()
+		if err != nil {
+			return status.Error(codes.Internal, "Failed to restart syslog service")
```
Comment (Dmaddu) [inline components/automate-deployment/pkg/server/automate_config.go:270]: same thing here as well.

File: components/automate-deployment/pkg/server/automate_config.go [hunk 4]
```diff
@@ -235,3 +254,139 @@ func (s *server) updateUserOverrideConfigFromRestoreBackupRequest(req *api.Resto
...
+			return err
+		}
+	}
+
+	//Rollback the config if requested
+	if req.GetConfig().GetGlobal().GetV1().GetLog().GetRedirectSysLog().GetValue() == false &&
+		existingCopy.GetGlobal().GetV1().GetLog().GetRedirectSysLog().GetValue() == true {
+		err := removeConfigFileForAutomateSyslog()
+		if err != nil {
+			return status.Error(codes.Internal, "Failed to remove configuration into syslog")
```
Comment (Dmaddu) [inline components/automate-deployment/pkg/server/automate_config.go:284]: same thing here as well.

File: components/automate-deployment/pkg/server/automate_config.go [hunk 5]
```diff
@@ -235,3 +254,139 @@ func (s *server) updateUserOverrideConfigFromRestoreBackupRequest(req *api.Resto
...
+		}
+		err = restartSyslogService()
+		if err != nil {
+			return status.Error(codes.Internal, "Failed to restart syslog service while removing config file for syslog")
```
Comment (Dmaddu) [inline components/automate-deployment/pkg/server/automate_config.go:288]: same thing here as well.

File: components/automate-deployment/pkg/server/automate_config.go [hunk 6]
```diff
@@ -235,3 +254,139 @@ func (s *server) updateUserOverrideConfigFromRestoreBackupRequest(req *api.Resto
...
+			logrus.Errorf("cannot rollback logrotate: %v", err)
+			return err
+		}
+
+	}
+
+	return nil
+}
+
+func restartSyslogService() error {
```
Comment (Dmaddu) [inline components/automate-deployment/pkg/server/automate_config.go:301]: please add a description to all the new methods, at least one or two liners describe the actual operation we are doing over that method.

**Review by Dmaddu — CHANGES_REQUESTED** [review]

File: components/automate-deployment/pkg/server/helpers.go [hunk 7]
```diff
@@ -0,0 +1,27 @@
+package server
+
+import (
+	"bytes"
+	"fmt"
+)
+
+func LogRotateConf(path string, params ...string) string {
```
Comment (Dmaddu) [inline components/automate-deployment/pkg/server/helpers.go:8]: what if we pass empty params? something like 
LogRotateConf(" some path", "copytruncate", "")

**Review by Dmaddu — APPROVED** [review]

**Pull request merged**